### ISSUE TITLE
How to split dataset by multi workers?

### ISSUE THREAD

**Issue by anpark** [horovod/horovod#249]
HI, if i have 5 parts input dataset in hdfs, then if i use 5 workers to train 2 epochs
i think worker 0 read part-0 2epochs, worker 1 read part-1 2 epochs,..... is it right, if yes, how to do that by horovod?

**Comment by alsrgv** [comment]
If your # of files exactly matches a number of workers or is a multiple of it, you could use [tf.train.match_filenames_once](https://www.tensorflow.org/api_docs/python/tf/train/match_filenames_once) with `hvd.rank()` as part of file name pattern.

That said, it would be better and more scalable to create # of training files in HDFS which is a 10+ multiple of # of workers (e.g. 50 files for 5 workers) and then just have every worker randomly read files / records from the dataset for X iterations, where `X = <total # of examples in dataset> / <batch size> / <number of workers>`.

**Comment by anpark** [comment]
but i don't see that code in tf-benchmark, how do you test multi workers?

**Comment by alsrgv** [comment]
TF benchmark does random file reading via different [random seed](https://github.com/tensorflow/benchmarks/blob/master/scripts/tf_cnn_benchmarks/benchmark_cnn.py#L1688) on different workers.

Here's an example of per-epoch adjustment of number of steps: https://github.com/uber/horovod/blob/master/examples/keras_imagenet_resnet50.py#L140

**Comment by anpark** [comment]
but that's not exact shard, maybe have same file part on different workers, i'm not sure that's ok?

**Comment by anpark** [comment]
because i found that horovod runs much faster then PS with sync, but converged much worse than parameter server

**Comment by alsrgv** [comment]
We did an experiment with ImageNet where we trained one model with random reading, which could indeed face issues with multiple workers reading the same shard on certain occasions, and also trained model with pre-determined partitions that are re-shuffled every epoch.  Both experiments converged to the same accuracy.

**Comment by alsrgv** [comment]
That's an interesting observation.  Can you share details of your experiment?  Note - in ImageNet experiment # of file shards was 10x of # of workers.

**Comment by anpark** [comment]
I ran the benchmark for cifar10 in tensorflow/models  using 1 node with 4 gpus:
accuracy: 0.9114 by PS sync or official multi gpu, time: 2 hours, but 0.8931 by hornovd, time: 30min

model: resnet32_v2
Batch size: 128 global 32 per device 
num_samples:50000, num_epochs:250.00, num_steps:97657 
piecewise_learning_rate_schedule: 0.1;100;0.01;150;0.001;200;0.0001
data_format:NCHW

**Comment by alsrgv** [comment]
@anpark, I've trained a couple of models and got approximately the same result between single-GPU (P100 with batch size 128) and 4xGPU with batch size 32 each.

Single GPU:
```
$ python tf_cnn_benchmarks.py --model resnet32_v2 --data_name cifar10 --batch_size 128 --data_dir ~/cifar-10-batches-py --piecewise_learning_rate_schedule '0.1;100;0.01;150;0.001;200;0.0001' --num_epochs 250 --train_dir ~/cifar-single-gpu --save_model_secs 60 --save_summaries_steps 1000
...
$ python tf_cnn_benchmarks.py --eval --model resnet32_v2 --data_name cifar10 --batch_size 128 --num_epochs 1 --data_dir ~/cifar-10-batches-py --train_dir ~/cifar-single-gpu
Accuracy @ 1 = 0.9011 Accuracy @ 5 = 0.9962 [9984 examples]
```

4xGPU with Horovod:
```
$ mpirun -np 4 python tf_cnn_benchmarks.py --model resnet32_v2 --data_name cifar10 --batch_size 32 --data_dir ~/cifar-10-batches-py --piecewise_learning_rate_schedule '0.1;100;0.01;150;0.001;200;0.0001' --num_epochs 250 --train_dir ~/cifar-horovod --save_model_secs 60 --save_summaries_steps 1000 --variable_update horovod
...
$ python tf_cnn_benchmarks.py --eval --model resnet32_v2 --data_name cifar10 --batch_size 32 --num_epochs 1 --data_dir ~/cifar-10-batches-py --train_dir ~/cifar-horovod/ --variable_update horovod
Accuracy @ 1 = 0.8994 Accuracy @ 5 = 0.9963 [9984 examples]
```

**Comment by anpark** [comment]
@alsrgv Thanks!
I found the problem is my env is out of date, now it's ok
Horovod cost 50min, PS sync and multi-gpu using 1hour, Horovod is a Great Job!

**Comment by sudheerExperiments** [comment]
@alsrgv, sure I will try it out and will let you know how it goes.

**Comment by alsrgv** [comment]
@sudheerExperiments, can you try checking out a new copy of https://github.com/tensorflow/benchmarks?  I can successfully checkout or reset to `4d8aa531132d5b144750c1a48c7c8a29cc0eba6a:  HEAD is now at 4d8aa53 Allow --gradient_repacking, --compact_gradient_transfer, and --variable_consistency to work independently of --hierarchical_copy and each other, Part 1.`

**Comment by sudheerExperiments** [comment]
Thanks @alsrgv, its working now. 

>**Note:** To people who come across this issue in future. Please do not reinitialize empty git repository. Just clone and reset to the commit id given by alex. This will rollback git repository to run benchmarks with 'tensorflow 12.0', 'cuda 9.2', and 'cudnn 7'.
